Add psalm static analysis, improve type inference

tux-rampage opened the pull request:
<!--
Fill in the relevant information below to help triage your issue.

Pick the target branch based on the following criteria:
  * Documentation improvement: master branch
  * Bugfix: master branch
  * QA improvement (additional tests, CS fixes, etc.) that does not change code
    behavior: master branch
  * New feature, or refactor of existing code: develop branch

You MUST provide a signoff in your commits for us to be able to accept your
patch; you can do this by providing either the --signoff or -s flag when using
"git commit". Please see the project contributing guide and
https://developercertificate.org for details.
-->

|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | yes
| RFC           | no
| QA            | yes

### Description

As decided by the technical steering committee, components should be covered by psalm static analysis.
This PR attempts to replace phpstan with psalm.

It will also fix reported issues and provide support for templates/generics when possible.

<!--
Tell us about why this change is necessary:
- Are you fixing a bug or providing a failing unit test to demonstrate a bug?
  - How do you reproduce it?
  - What did you expect to happen?
  - What actually happened?
  - TARGET THE master BRANCH

- Are you adding documentation?
  - TARGET THE master BRANCH

- Are you providing a QA improvement (additional tests, CS fixes, etc.) that
  does not change behavior?
  - Explain why the changes are necessary
  - TARGET THE master BRANCH

- Are you fixing a BC Break?
  - How do you reproduce it?
  - What was the previous behavior?
  - What is the current behavior?
  - TARGET THE master BRANCH

- Are you adding something the library currently does not support?
  - Why should it be added?
  - What will it enable?
  - How will the code be used?
  - TARGET THE develop BRANCH

- Are you refactoring code?
  - Why do you feel the refactor is necessary?
  - What types of refactoring are you doing?
  - TARGET THE develop BRANCH
-->

tux-rampage committed:
Remove phpstan

Signed-off-by: tux-rampage <tuxrampage@gmail.com>

tux-rampage committed:
Add psalm

Signed-off-by: tux-rampage <tuxrampage@gmail.com>

tux-rampage commented:
Note: Psalm cannot be installed as dependency with php 8 at the moment. See vimeo/psalm#4408

boesing commented on psalm.xml:
```diff
@@ -0,0 +1,39 @@
+        </ignoreFiles>
+    </projectFiles>
+
+    <issueHandlers>
```
We should generate a baseline instead of suppressing issues by default

tux-rampage replied:
@boesing This is from the suggested template: https://github.com/mezzio/repo-template/blob/6124090610ef1c6d6b0c80a063cafacc3ec68f38/psalm.xml.dist

Not sure if needed when the phpunit plugin for psalm is used. I guess we can drop them.

I'm not done yet, my goal is to not need a baseline.

boesing replied:
Sounds good. I've created that issue template before I've implemented psalm to any package, so yes, sadly its quite outdated. However, thanks for being so passionate :-)

tux-rampage replied:
While working on this, it would require some major rewrites, especially the config handling.
So I'll fix as much as possible but try to keep the changeset low and target the rest with separate PR's

Refactoring the config handling from arrays to types is on my tasklist anyways but out of scope for this PR.

tux-rampage replied:
Oh and getting rid of the baseline in the end is still the final goal ;-)

weierophinney commented:
@tux-rampage I've rebased your branch based on current 3.3.x, which includes the switch from Travis to GHA for CI purposes. I've also renamed the `psalm.xml` file to `psalm.xml.dist` (for consistency with other repos) and ensured an up-to-date `composer.lock` is available (built using PHP 7.4).

CS checks and Psalm checks are currently failing, which you can now see in the checks list.

tux-rampage committed:
Fix coding style

Signed-off-by: tux-rampage <tuxrampage@gmail.com>

tux-rampage committed:
Add psalm baseline

Signed-off-by: tux-rampage <tuxrampage@gmail.com>

tux-rampage commented on composer.json:
```diff
@@ -26,7 +26,7 @@
     },
     "require": {
-        "php": "^7.3 || ~8.0.0 || ~8.1.0",
+        "php": "^7.4 || ~8.0.0 || ~8.1.0",
```
PHP 7.3 runs out of support this year. Frankly I'd drop PHP 7 entirely here and push to 8.0 and 8.1 (separate PR)

tux-rampage commented on phpcs.xml:
```diff
@@ -22,10 +22,16 @@
 
+    <!-- This must be excluded since it is not compatible with php < 8 -->
+    <rule ref="SlevomatCodingStandard.Classes.ModernClassNameReference.ClassNameReferencedViaFunctionCal">
+        <exclude-pattern>*</exclude-pattern>
```
Since we still have to support 7.4 this rule must be excluded or test cases will fail.

`$object::class` does not work with php < 8 and tests will fail.

Ocramius replied:
Let's add  `<config name="php_version" value="70499"/>` at the top instead

tux-rampage replied:
Perfect, that's way better than my exclude.

tux-rampage commented on phpunit.xml.dist:
```diff
@@ -1,5 +1,9 @@
+<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:noNamespaceSchemaLocation="https://schema.phpunit.de/9.5/phpunit.xsd"
+         bootstrap="vendor/autoload.php"
+         convertDeprecationsToExceptions="true"
```
With the most recent PHPUnit update, this must be enabled explicitly.
Is runtime deprecations still a thing? Can we drop them in favour of static analysis tools like psalm.

There are good reasons for not triggering deprecations at runtime. They're also a horror to maintain.

See #30 and #31

tux-rampage commented on src/Config.php:
```diff
@@ -192,15 +212,18 @@ public function isAlias(string $name): bool
     public function getConfiguredTypeNames(): array
     {
-        return array_keys($this->types);
+        return array_map('strval', array_keys($this->types));
```
Make sure this is actually a string list. PHP always converts integerish strings to an `int` when they are used as array key, so this is actually required: https://3v4l.org/O2Wd9

tux-rampage commented on src/Config.php:
```diff
@@ -192,15 +212,18 @@ public function isAlias(string $name): bool
     {
         if ($context) {
+            /** @psalm-suppress MixedArrayAssignment TODO: Eliminate array structures with the next releases */
             $this->types[$context]['preferences'][$type] = $preference;
```
This is something I'd like to eliminate completely in future versions of this component.
All structures should be explicitly shaped and typed. Config elements should be immutable value objects.

tux-rampage commented on src/ConfigProvider.php:
```diff
@@ -27,15 +29,17 @@ public function __invoke(): array
-                LegacyInjectorGenerator::class => CodeGenerator\InjectorGenerator::class,
+                'Zend\Di\InjectorInterface'               => InjectorInterface::class,
+                'Zend\Di\ConfigInterface'                 => ConfigInterface::class,
+                'Zend\Di\CodeGenerator\InjectorGenerator' => CodeGenerator\InjectorGenerator::class,
```
Is this still a thing? Could we drop this since this package is now marked as conflicting to `zend-di`?

Ocramius replied:
Not really: these string keys were still strings, and not class aliases

tux-rampage replied:
Do you have a good idea on how to get rid of this gracefully?

tux-rampage commented on test/ConfigTest.php:
```diff
@@ -12,104 +12,116 @@
     {
         $config = new Config($this->fixture);
-        $this->assertEquals(['Foo', 'Bar'], $config->getConfiguredTypeNames());
+        $this->assertEquals([TestAsset\Config\SomeClass::class, 'SomeAlias'], $config->getConfiguredTypeNames());
```
All changes like this were necessary since we do now type with `class-string` explicitly. This was caused especially by the `typeOf` config key which always had to be an existing interface or class name (See [docs](https://docs.laminas.dev/laminas-di/v3/config/)).

It just revealed a logical bug that slipped due to lax typing.

tux-rampage reviewed (commented):
I have added as types as explicit as possible for now. I have decided to not work through all tests since this would make this already big PR even bigger and harder to review. Test-Code issues are added to the baseline. We can keep up with smaller PR to work these things out.

Ocramius commented on src/CodeGenerator/FactoryInterface.php:
```diff
@@ -6,12 +6,16 @@
 use Psr\Container\ContainerInterface;
 
+/**
+ * @template T extends object
```
Will this work in inheritance?

tux-rampage replied:
Yepp should be no issue. The chance the this is implemented directly is almost zero.

tux-rampage replied:
Oh yes, Implementing classes should use `@template-implements`. I adjusted the tests for the generated classes.

Edit: See https://github.com/laminas/laminas-di/pull/21#discussion_r717930889

boesing replied:
Is this a working psalm annotation?
Shouldn't this be `@template T of object` instead?

tux-rampage replied:
Correct, ith schould be `of`. This was TypeScript comming through.

Ocramius commented on src/CodeGenerator/GeneratorTrait.php:
```diff
@@ -44,6 +49,8 @@ protected function ensureDirectory(string $dir)
      * @throws LogicException
      * @throws GenerateCodeException
+     * @return void
+     * @psalm-assert non-empty-string $this->outputDirectory
```
Not sure this works with private state :thinking:

tux-rampage replied:
Yes it does

tux-rampage replied:
Here you go: https://psalm.dev/r/205038dd2c

Ocramius commented on src/Container/AutowireFactory.php:
```diff
@@ -36,7 +37,7 @@ private function getInjector(ContainerInterface $container)
      * Check creatability of the requested name
      *
-     * @param string $requestedName
+     * @param string|Stringable $requestedName
```
Oooof, no, we should not really accept `Stringable` :-\

tux-rampage replied:
That was just to keep BC with the config arrays. But I get your point.

tux-rampage replied:
Nope you're right, this is actually extending the method signature which was not my intention.
Could we place a native `string` typehint? 

Some may argue that this would somehow be a BC break for user's who ignore the DocBlock and rely on the string cast behavior of this method.

Adding a native `string` type hint is valid imo, since everything else already violates the existing contract from the DocBlock.

tux-rampage committed:
Add min php version to cs config

Signed-off-by: tux-rampage <tuxrampage@gmail.com>

tux-rampage committed:
Remove incorrect Stringable from PhpDoc

Signed-off-by: tux-rampage <tuxrampage@gmail.com>

tux-rampage committed:
Add template implements for generated factory classes

Signed-off-by: tux-rampage <tuxrampage@gmail.com>

tux-rampage commented on test/_files/expected-codegen-results/factories/with-params.php:
```diff
@@ -28,10 +31,15 @@ public function create(ContainerInterface $container, array $options = [])
             ];
 
+        /** @psalm-suppress MixedArgument */
         return new \LaminasTest\Di\TestAsset\Constructor\MixedArguments(...$args);
```
The interface type would correctly be: `public function create(ContainerInterface $container, PartialArray<ConstructArgsArray<T>> $options = []): T`

This can only be achieved with a psalm plug-in as far as I know. But this is out of scope of this PR.
To avoid additional overhead in these generated classes, a suppress is added and responsibility is passed to PHP's type system until a better solution can be provided.

tux-rampage commented:
Thanks for your reviews @boesing and @Ocramius
I've addressed your feedback. Can you please re-check?

boesing commented on phpcs.xml:
```diff
@@ -1,16 +1,17 @@
 <?xml version="1.0"?>
 <ruleset xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="vendor/squizlabs/php_codesniffer/phpcs.xsd">
 
+    <config name="php_version" value="70499"/>
```
this shouldn't be necessary as we've changed the CI matrix so that QA tooling (such as codestyle checks) are only executed on the minimum supported PHP version of a component.

As this component requires PHP 7.4 as its minimum, you can remove this line here so it won't lead to issues when removing PHP 7.4.

tux-rampage replied:
I see. I guess this happened since I'm running php 8 locally and stumbled upon this.

Ocramius replied:
Kept this: still useful for local runs

boesing commented on psalm.xml.dist:
```diff
@@ -0,0 +1,35 @@
+    </issueHandlers>
+
+    <stubs>
+        <file name="./stubs/psr-container.phpstub" />
```
I prefer using `lctrs/psalm-psr-container-plugin` but I know that this is a bit opinionated and thus, I know that @Ocramius does not like such magic.

But I'd suggest we go either with the existing psalm-plugin or not.
In both cases, we should drop the stub from this project.

tux-rampage replied:
And I agree with @Ocramius since all that Plugin does (as far as I got it) is re-implementing what psalm already can do with templates/generics (and therefore adds a greater potential to break or prevent a psalm upgrade).

Anyways I can replace it with that Plugin. Dropping this without a replacement would require additional typecheck code and make this already big PR even bigger.

Edit: Submitted this comment too soon.

boesing commented on src/CodeGenerator/AbstractInjector.php:
```diff
@@ -13,10 +13,10 @@
 abstract class AbstractInjector implements InjectorInterface
 {
-    /** @var string[]|FactoryInterface[] */
+    /** @var array<string, class-string<FactoryInterface>|FactoryInterface> */
```
*maybe* bump minimum requirement of `laminas-servicemanager`.
I've added plenty of psalm types there which can be imported here for re-usage.

tux-rampage replied:
laminas-servicemanager is a dev-dependency and it's not meant to be a requirement. 
Importing types from this project would require to make it a runtime dependency just to pass static analysis, which is too much imho.

Ocramius replied:
This is tested against a recent and locked servicemanager: all good

boesing commented on src/CodeGenerator/AutoloadGenerator.php:
```diff
@@ -49,6 +49,7 @@ private function buildFromTemplate(string $templateFile, string $outputFile, arr
         $template = file_get_contents($templateFile);
 
         assert(is_string($template));
+        assert(is_string($this->outputDirectory));
```
Imho, thats not safe. This is only true in case someone called the setter. Using php `assert` in this case is *not* how it should be used.

https://github.com/laminas/technical-steering-committee/blob/main/meetings/minutes/2021-08-02-TSC-Minutes.md#allow-php-assert-along-with-webmozart-assert-in-specific-cases

tux-rampage replied:
You are right. `$this->ensureOutputDirectory()` would be correct here, which does provide the correct assert and the required directory existence checks / create attempts.

boesing commented on src/CodeGenerator/FactoryGenerator.php:
```diff
@@ -183,6 +183,8 @@ public function generate(string $class): string
         $factoryClassName                          = $this->namespace . '\\' . $this->buildClassName($class);
         [$namespace, $unqualifiedFactoryClassName] = $this->splitFullyQualifiedClassName($factoryClassName);
 
+        assert(is_string($this->outputDirectory));
```
Imho, thats not safe. This is only true in case someone called the setter. Using php assert in this case is not how it should be used.

https://github.com/laminas/technical-steering-committee/blob/main/meetings/minutes/2021-08-02-TSC-Minutes.md#allow-php-assert-along-with-webmozart-assert-in-specific-cases

tux-rampage replied:
Correct. Same as for the `AutoloadGenerator`.

boesing commented on src/CodeGenerator/InjectorGenerator.php:
```diff
@@ -98,6 +98,8 @@ private function buildFromTemplate(string $templateFile, string $outputFile, arr
 
     private function generateInjector(): void
     {
+        assert(is_string($this->outputDirectory));
```
Imho, thats not safe. This is only true in case someone called the setter. Using php assert in this case is not how it should be used.

https://github.com/laminas/technical-steering-committee/blob/main/meetings/minutes/2021-08-02-TSC-Minutes.md#allow-php-assert-along-with-webmozart-assert-in-specific-cases

boesing commented on src/Config.php:
```diff
@@ -71,50 +74,62 @@
+ * @see \Laminas\Di\Resolver\TypeInjection  A container to force looking up a specific type instance for injection
+ *
+ * @psalm-type TypeConfigArray = array{
+ *  typeOf?: class-string|null,
```
Do we actually allow `null` or is that something like a real-world example?

I would say, if `typeOf` is provided (thanks to `?`, it is not a required array key), only `class-string` should be provided. 🤷🏼‍♂️

tux-rampage replied:
That depends. When we want to provide a way to remove existing settings from a ConfigProvider or project level config file, then allowing `null` is the only option, since the Laminas `config` array is merged from config providers or MVC modules.

I am not sure if this method this is actively used, but it is a valid possibility at the moment.

boesing commented on src/Config.php:
```diff
@@ -71,50 +74,62 @@
+ *
+ * @psalm-type TypeConfigArray = array{
+ *  typeOf?: class-string|null,
+ *  preferences?: array<string, string>|null,
```
Do we actually allow `null` or is that something like a real-world example?

I would say, if `preferences` is provided (thanks to ?, it is not a required array key), only a map of strings should be provided. 🤷🏼‍♂️

tux-rampage replied:
See `typeOf` remarks

boesing commented on src/Config.php:
```diff
@@ -71,50 +74,62 @@
+ * @psalm-type TypeConfigArray = array{
+ *  typeOf?: class-string|null,
+ *  preferences?: array<string, string>|null,
+ *  parameters?: array<string, mixed>|null
```
Do we actually allow `null` or is that something like a real-world example?

I would say, if `parameters` is provided (thanks to ?, it is not a required array key), only a map of strings should be provided. 🤷🏼‍♂️

tux-rampage replied:
See `typeOf` remarks

boesing commented on src/Container/AutowireFactory.php:
```diff
@@ -17,13 +17,13 @@ class AutowireFactory
     {
         $injector = $container->get(InjectorInterface::class);
 
+        /** @psalm-suppress DocblockTypeContradiction Can we remove all code below and return directly from container get? */
```
As mentioned in another comment, just because you request `InjectorInterface::class` from the container, that **must** not be an instance of the requested class-string.

I would be fine with that assertion and also with crashing the application with a type-error if some upstream dev thinks he can hassle with the dependency injection but thats just my opinion about this but @Ocramius had a point on this.

tux-rampage replied:
IMHO: We can assume that the container provides matching class instances, since this is what laminas-di provides: Autowire for Psr Contrainers when requesting `$container->get(Some::class)`.

I'd also argue that this can be handled by PHP's type checker rather than checking it manually - in the end it's less code to maintain.

boesing commented on src/Container/ConfigFactory.php:
```diff
@@ -20,21 +24,32 @@
      */
     public function create(ContainerInterface $container): ConfigInterface
     {
+        /** @var mixed $config */
```
`array<string,mixed>|ArrayAccess<string,mixed>` would be a better annotation imho

tux-rampage replied:
> As mentioned in another comment, just because you request `InjectorInterface::class` from the container, that **must** not be an instance of the requested class-string.

This can be argued here as well. This code is not even requesting a concrete class but a `string` name without any type relation from the container which could be anything. Especially since laminas-di is a component that does not require a laminas/mezzio project.

boesing commented on src/Container/GeneratorFactory.php:
```diff
@@ -19,13 +22,18 @@ private function getConfig(ContainerInterface $container): ConfigInterface
-        if ($container->has(LegacyConfigInterace::class)) {
-            return $container->get(LegacyConfigInterace::class);
+        if ($container->has('Zend\Di\ConfigInterface')) {
+            /** @psalm-var ConfigInterface */
```
Huh, does this really provide any value to psalm? There is no var which is being assigned 🤔

tux-rampage replied:
That's a way to tell psalm the resulting type of the (rhs) expression that follows this comment. It does not require a variable name, so it also can be placed inline:
https://psalm.dev/docs/annotating_code/supported_annotations/#off-label-usage-of-the-var-tag

boesing commented on src/Definition/Reflection/ClassDefinition.php:
```diff
@@ -69,7 +71,10 @@ public function getInterfaces(): array
 
-    private function reflectParameters()
+    /**
+     * @psalm-assert array<string, Parameter> $this->parameters
```
instead of this, wouldn't it be better to simply set `parameters` to array with a default value?

tux-rampage replied:
Probably yes, but this would require additional refactoring to this class since `null` is used by getter methods. I just added the types to reflect the existing behaviour.

boesing commented on src/Definition/Reflection/ClassDefinition.php:
```diff
@@ -34,7 +33,10 @@ public function __construct($class)
 
-    private function reflectSupertypes()
+    /**
+     * @psalm-assert list<string> $this->supertypes
```
instead of this, wouldn't it be better to simply set `supertypes` to array with a default value?

tux-rampage replied:
Probably yes, but this would require additional refactoring to this class since `null` is used by getter methods. I just added the types to reflect the existing behaviour.

boesing commented on src/Definition/Reflection/ClassDefinition.php:
```diff
@@ -16,14 +15,14 @@ class ClassDefinition implements ClassDefinitionInterface
     private $parameters;
 
-    /** @var string[] */
+    /** @var list<string>|null */
```
if setting an empty array as default, we could avoid having `null` here imho

tux-rampage replied:
Probably yes, but this would require additional refactoring to this class since `null` is used by getter methods. I just added the types to reflect the existing behaviour.

boesing commented on src/Definition/Reflection/ClassDefinition.php:
```diff
@@ -16,14 +15,14 @@ class ClassDefinition implements ClassDefinitionInterface
     private $reflection;
 
-    /** @var Parameter[] */
+    /** @var Parameter[]|null */
```
if setting an empty array as default, we could avoid having `null` here imho

tux-rampage replied:
Probably yes, but this would require additional refactoring to this class since `null` is used by getter methods. I just added the types to reflect the existing behaviour.

boesing commented on src/Definition/RuntimeDefinition.php:
```diff
@@ -16,10 +16,10 @@
     private $definition = [];
 
-    /** @var bool[] */
+    /** @var array<string, bool>|null */
```
if setting an empty array as default, we could avoid having `null` here imho

tux-rampage replied:
Probably yes, but this would require additional refactoring to this class since `null` is used by getter methods. I just added the types to reflect the existing behaviour.

boesing reviewed (changes requested):
Ooof, well this PR has really plenty of code changes and its not that easy to review. However, I've added a few things which should be addressed.

Especially those `null` properties which could be empty arrays to make things simpler.

tux-rampage commented:
@boesing Thanks for taking time to review,

> Ooof, well this PR has really plenty of code changes and its not that easy to review. However, I've added a few things which should be addressed.

Yes, that's why I focused on adding types for the existing implementation without refactoring where possible. You pointed out a lot of good and valid points to improve.

I'd prefer to address the Improvements in separate smaller PRs and focus on the bugs you pointed out (like the incorrect string-asserts). As you noted, this one is already hard enough to review, and I don't want to add more complexity here as necessary. Would this be OK for you?

Ocramius commented:
@tux-rampage a couple files require a rebase - nothing major, as we mostly only removed PHP 7.3 and locked everything with `composer.lock`.

I suggest moving most issues to Psalm's baseline and "closing the chapter" meanwhile :)

tux-rampage commented:
> @tux-rampage a couple files require a rebase - nothing major, as we mostly only removed PHP 7.3 and locked everything with `composer.lock`.
> 
> I suggest moving most issues to Psalm's baseline and "closing the chapter" meanwhile :)

Perfect agreed. I'd rather like to do this in smaller PRs that are easier to review.

Ocramius committed:
Removed reference to dropped prophecy class (leading to test failures)

Ocramius committed:
Removing intersection type scan from Psalm's watchful eye

`vimeo/psalm:^4` does not support native intersection types yet: `vimeo/psalm:^5` will do,
but for now, we just exclude this code path from analysis.

Ocramius committed:
More precise psalm config: stricter, finding unused code, finding unused suppressions

Ocramius committed:
Fixing or baselining most reported psalm errors

Note: many reflection-related suppressions depend on https://github.com/vimeo/psalm/pull/8722

Also, we preserved many assertions on `class-string`, since removing
input validation would be a BC break.

Ocramius committed:
Re-generated psalm baseline

Ocramius committed:
Removed unnecessary inline var annotations, demonstrating type inference improvements

Ocramius committed:
Applied automated CS fixes

Ocramius committed:
Ignore analysis of symbols containing intersection types

Ocramius committed:
Trick PHPCS into skipping analysis for a class containing PHP 8.1 intersection types

Ocramius commented:
After rebasing and many tiny adjustments (Psalm became **MUCH** more powerful over the last whole year, I decided that this is good to go as-is.

Thanks @tux-rampage, and sorry for dragging this on: very valuable work!

tux-rampage commented:
> After rebasing and many tiny adjustments (Psalm became **MUCH** more powerful over the last whole year, I decided that this is good to go as-is.
> 
> Thanks @tux-rampage, and sorry for dragging this on: very valuable work!

Thank you for finishing this. :+1:
